WTF: Use __bridge_retained to help the compiler keep track of CF objects https://bugs.webkit.org/show_bug.cgi?id=253820

LumioseSil opened the pull request:
#### b6a50d258a8e656e38e2d0b92b8cfa96eb3bc692
<pre>
WTF: Use __bridge_retained to help the compiler keep track of CF objects <a href="https://bugs.webkit.org/show_bug.cgi?id=253820">https://bugs.webkit.org/show_bug.cgi?id=253820</a>

Reviewed by Darin Adler.

As we use __bridged_release when converting to NS Objects,
it makes sense to do the inverse when converting to CF Objects.

* Source/WTF/wtf/RetainPtr.h: Statically assert that adoptNS is only used
  for NS objects.

* Source/WTF/wtf/cocoa/TypeCastsCocoa.h: Use __bridge_retained to
  convert NS objects to CF objects.

Canonical link: <a href="https://commits.webkit.org/261860@main">https://commits.webkit.org/261860@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ed8097549038c3fb9e7d0f24d972d1d9f20b7b9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/113082 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22215 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1748 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4853 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121545 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23585 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13398 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/6089 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118869 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/17525 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100788 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/106164 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99479 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1324 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46553 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/101317 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14517 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1365 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/12674 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15232 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/10700 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/102874 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20526 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53357 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/32098 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8279 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/17078 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/110927 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/27399 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/RetainPtr.h:
```diff
@@ -323,6 +323,7 @@ template<typename T> constexpr RetainPtr<T> adoptCF(T CF_RELEASES_ARGUMENT ptr)
 #ifdef __OBJC__
 template<typename T> inline RetainPtr<typename RetainPtr<T>::HelperPtrType> adoptNS(T NS_RELEASES_ARGUMENT ptr)
 {
+    static_assert((std::is_convertible<T, id>::value), "Don't use adoptNS with Core Foundation pointer types, use adoptCF.");
```
Please use `std::is_convertible_v` and also omit unneeded extra set of parentheses.

AZero13 replied:
I will do you one better, I will change all the ::value to _v

darinadler reviewed (approved):
Looks good.

darinadler reviewed (approved):
Not a huge fan of the change log comment; this doesn’t need to be fixed before landing but here are my thoughts for the record.

I believe “very hacky” is an incorrect way to describe what’s wrong with the code we are replacing. The personification of the compiler saying it “probably doesn’t” have a “strong idea” is not helpful; if there’s a concrete way this change helps the compiler we could state that. Expressing an imprecise implication that is possibly inaccurate doesn’t help the reader’s understanding. The comment says that `__bridge_retained` is a “macro” — it is not. The paths here use Windows-style backslashes, not Unix-style forward slashes. Our change log style is the Unix one.

Using precise and direct language in the change log is helpful to project contributors long term; we should strive to do better.

On the other hand, since `__bridge_retained` is available, and does the right thing here, it’s **great** to use it. Maybe someday we will discover a concrete benefit, perhaps the static analyzer will work better and catch some kinds of mistakes. And it’s nice to add the `static_assert` to `adoptNS`. Glad to see that there’s no existing mistake in the code that it uncovered.

LumioseSil committed:
WTF: Use __bridge_retained to help the compiler keep track of CF objects https://bugs.webkit.org/show_bug.cgi?id=253820

Reviewed by Darin Adler.

As we use __bridged_release when converting to NS Objects,
it makes sense to do the inverse when converting to CF Objects.

* Source/WTF/wtf/RetainPtr.h: Statically assert that adoptNS is only used
  for NS objects.

* Source/WTF/wtf/cocoa/TypeCastsCocoa.h: Use __bridge_retained to
  convert NS objects to CF objects.

Canonical link: https://commits.webkit.org/261860@main
